build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1235,6 +1235,33 @@ message OutputDirectory {
   // [Tree][build.bazel.remote.execution.v2.Tree] proto containing the
   // directory's contents.
   Digest tree_digest = 3;
+
+  // If set, consumers MAY make the following assumptions about the
+  // directories contained in the the Tree, so that it may be
+  // instantiated on a local file system by scanning through it
+  // sequentially:
+  //
+  // - All directories with the same binary representation are stored
+  //   exactly once.
+  // - All directories, apart from the root directory, are referenced by
+  //   at least one parent directory.
+  // - Directories are stored in topological order, with parents being
+  //   stored before the child. The root directory is thus the first to
+  //   be stored.
+  //
+  // Additionally, the Tree MUST be encoded as a stream of records,
+  // where each record has the following format:
+  //
+  // - A single tag byte, having one of the following values:
+  //   - First record (the root directory): 0x0a
+  //   - Any subsequent records (child directories): 0x12
+  // - The size of the directory, encoded as a base 128 varint.
+  // - The contents of the directory, encoded as a binary serialized
+  //   Protobuf message.
```

Comment:
Do all these requirements come from the proto3 encoding?

Reply:
> Do all these requirements come from the proto3 encoding?

Yes, but they have been strengthened. For example, Protobuf field tags are also encoded as base 128 varints, meaning that you could also write tag 0x0a (for the root directory) as one of the following:

- 0x8a 0x00
- 0x8a 0x80 0x00
- 0x8a 0x80 0x80 0x00
- 0x8a 0x80 0x80 0x80 0x00
- 0x8a 0x80 0x80 0x80 0x80 0x80 0x80 0x80 ... 0x80 0x00

Also, plain Protobuf would permit placing the root directory anywhere you want; it doesn't need to come before the child directories. It may even be placed in between two of them. Both of these things are of course completely undesirable, which is why I decided to give a complete description of what the binary format looks like.

> I think it's worth documenting where these values come from.

Yes! I've just made two improvements:

1. I've added derivations of where 0x0a and 0x12 come from. They are basically `(x << 3) | 2`, where `x` is the field number.
2. I've added a link to the Protobuf encoding format.